hooks/react-client/useNavigationState.ts

```diff
@@ -0,0 +1,40 @@
+'use client';
+
+import type { RefObject } from 'react';
+import { useContext, useEffect } from 'react';
+
+import { NavigationStateContext } from '@/providers/navigationStateProvider';
+import { debounce } from '@/util/debounce';
+
+const useNavigationState = <T extends HTMLElement>(
+  id: string,
+  ref: RefObject<T>,
+  debounceTime = 300
+) => {
+  const navigationState = useContext(NavigationStateContext);
+
+  const handleScroll = debounce(() => {
+    if (ref.current) {
+      navigationState[id] = {
+        x: ref.current.scrollLeft,
+        y: ref.current.scrollTop,
+      };
+    }
+  }, debounceTime);
+
+  useEffect(() => {
+    const element = ref.current;
+    if (navigationState[id] && navigationState[id].y !== element?.scrollTop) {
+      element?.scroll({ top: navigationState[id].y, behavior: 'instant' });
+    }
+
+    element?.addEventListener('scroll', handleScroll, { passive: true });
```

Comment:
It's interesting that you use a scroll event listener. I have a different idea, what do you think about using a MutationObserver? https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver

Pretty much the idea is:

You have a utility that creates a mutation observer for an element (it can be the current element itself, by a ref, or something); The mutation observer would observe if the element got removed (which happens during navigation), and then you'd get the current scroll position of the component at the moment of unmounting.

Hence you won't need to rely on scroll event listeners. Just an alternative, both have the same effect IMO.

Reply:
That would be better. We don't need to track all scroll events (even if it is debounced). Only need the last scroll state the element had before it unmounted.

Reply:
I tried getting the scrollTop from layout effect on unmount and also MutationObserver. They both work but weirdly they don't work when we scroll to the bottom. It has something to do with the offsetTop. I made a [gist](https://gist.github.com/abizek/ae4232d6e683b16ffc64915fced5fe98) with the layout effect implementation.

Let's stick with the scroll event listener.

Reply:
Usage of `useLayoutEffect` should work nicely, what issues did you face by using an layout effect instead of a scroll listener?

Reply:
There is only the scroll-to-bottom issue. If the scrollTop can have a max value of 600 and offsetTop is 100, whenever we scroll above a scrollTop value of 500 (scrollTopMax - offsetTop), the value we get on unmount does not go above 500. Ideally, it should go up to 600. (Arbitrary values)

Reply:
I see, I think it makes sense for us to proceed with the scroll event and potentially iterate this further.

Thank you so much for all the work and time spent here, @abizek!